File: PlanetFinder.cs
// PlanetFinder.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace PlanetFinder
{
    class Program
    {
        static readonly Dictionary<string, double[]> PlanetData = new()
        {
            ["Mercury"] = new double[] { 252.2509, 4.09233445 },
            ["Venus"]   = new double[] { 181.9798, 1.60213043 },
            ["Earth"]   = new double[] { 100.4664, 0.98560911 },
            ["Mars"]    = new double[] { 355.4530, 0.52407178 },
            ["Jupiter"] = new double[] { 34.3515, 0.08309125 },
            ["Saturn"]  = new double[] { 50.0774, 0.03344482 },
            ["Uranus"]  = new double[] { 314.0550, 0.01172312 },
            ["Neptune"] = new double[] { 304.3487, 0.00597733 }
        };
        const double Obliquity = 23.4393;

        class Result
        {
            public string Planet { get; set; }
            public double RaDeg { get; set; }
            public double DecDeg { get; set; }
            public string Ra { get; set; }
            public string Dec { get; set; }
        }

        static void Main(string[] args)
        {
            var opts = ParseArgs(args);
            DateTime date;
            if (!string.IsNullOrEmpty(opts.Date))
            {
                if (!DateTime.TryParseExact(opts.Date, "yyyy-MM-dd", CultureInfo.InvariantCulture, DateTimeStyles.AssumeUniversal, out date))
                {
                    Console.WriteLine("\u001B[31mНеверный формат даты. Используйте YYYY-MM-DD\u001B[0m");
                    return;
                }
            }
            else
                date = DateTime.UtcNow;

            List<string> planets;
            if (!string.IsNullOrEmpty(opts.Planet))
            {
                if (!PlanetData.ContainsKey(opts.Planet))
                {
                    Console.WriteLine($"\u001B[31mПланета '{opts.Planet}' не найдена.\u001B[0m");
                    return;
                }
                planets = new List<string> { opts.Planet };
            }
            else
                planets = PlanetData.Keys.ToList();

            var results = new List<Result>();
            foreach (var p in planets)
            {
                var pos = ComputePosition(date, p);
                results.Add(new Result
                {
                    Planet = p,
                    RaDeg = pos.Item1,
                    DecDeg = pos.Item2,
                    Ra = RaToHms(pos.Item1),
                    Dec = DecToDms(pos.Item2)
                });
            }

            if (!string.IsNullOrEmpty(opts.ExportJson))
            {
                var json = JsonSerializer.Serialize(results, new JsonSerializerOptions { WriteIndented = true });
                File.WriteAllText(opts.ExportJson, json);
                Console.WriteLine($"\u001B[32mЭкспортировано в {opts.ExportJson} (JSON)\u001B[0m");
            }
            else if (!string.IsNullOrEmpty(opts.ExportCsv))
            {
                using var sw = new StreamWriter(opts.ExportCsv);
                sw.WriteLine("planet,ra_deg,dec_deg,ra_hms,dec_dms");
                foreach (var r in results)
                    sw.WriteLine($"{r.Planet},{r.RaDeg},{r.DecDeg},{r.Ra},{r.Dec}");
                Console.WriteLine($"\u001B[32mЭкспортировано в {opts.ExportCsv} (CSV)\u001B[0m");
            }
            else
            {
                Console.WriteLine($"\u001B[36mПоложение планет на {date:yyyy-MM-dd}:\u001B[0m\n");
                foreach (var r in results)
                {
                    Console.WriteLine($"\u001B[33m{r.Planet}:\u001B[0m");
                    Console.WriteLine($"  RA  = {r.Ra}  ({r.RaDeg:F2}°)");
                    Console.WriteLine($"  Dec = {r.Dec}  ({r.DecDeg:F2}°)");
                    Console.WriteLine();
                }
            }
        }

        static Options ParseArgs(string[] args)
        {
            var opts = new Options();
            for (int i = 0; i < args.Length; i++)
            {
                switch (args[i])
                {
                    case "--date": opts.Date = args[++i]; break;
                    case "--planet": opts.Planet = args[++i]; break;
                    case "--list": opts.List = true; break;
                    case "--export-json": opts.ExportJson = args[++i]; break;
                    case "--export-csv": opts.ExportCsv = args[++i]; break;
                }
            }
            return opts;
        }

        class Options
        {
            public string Date { get; set; }
            public string Planet { get; set; }
            public bool List { get; set; }
            public string ExportJson { get; set; }
            public string ExportCsv { get; set; }
        }

        static (double ra, double dec) EclipticToEquatorial(double lon, double lat, double obl)
        {
            double lonRad = lon * Math.PI / 180;
            double latRad = lat * Math.PI / 180;
            double oblRad = obl * Math.PI / 180;
            double ra = Math.Atan2(Math.Sin(lonRad) * Math.Cos(oblRad) - Math.Tan(latRad) * Math.Sin(oblRad),
                                   Math.Cos(lonRad));
            double dec = Math.Asin(Math.Sin(latRad) * Math.Cos(oblRad) + Math.Cos(latRad) * Math.Sin(oblRad) * Math.Sin(lonRad));
            double raDeg = (ra * 180 / Math.PI % 360 + 360) % 360;
            double decDeg = dec * 180 / Math.PI;
            return (raDeg, decDeg);
        }

        static string RaToHms(double raDeg)
        {
            double hours = raDeg / 15;
            int h = (int)hours;
            int m = (int)((hours - h) * 60);
            double s = (hours - h - m / 60.0) * 3600;
            return $"{h:00}h {m:00}m {s:00.00}s";
        }

        static string DecToDms(double decDeg)
        {
            char sign = decDeg >= 0 ? '+' : '-';
            decDeg = Math.Abs(decDeg);
            int d = (int)decDeg;
            int m = (int)((decDeg - d) * 60);
            double s = (decDeg - d - m / 60.0) * 3600;
            return $"{sign}{d:00}° {m:00}' {s:00.00}\"";
        }

        static (double ra, double dec) ComputePosition(DateTime date, string planet)
        {
            var j2000 = new DateTime(2000, 1, 1, 12, 0, 0, DateTimeKind.Utc);
            double days = (date.ToUniversalTime() - j2000).TotalDays;
            var data = PlanetData[planet];
            double lon = (data[0] + data[1] * days) % 360;
            return EclipticToEquatorial(lon, 0, Obliquity);
        }
    }
}

File: planet_finder.rs
// planet_finder.rs
use chrono::{DateTime, NaiveDate, TimeZone, Utc};
use clap::{App, Arg};
use serde::{Deserialize, Serialize};
use serde_json;
use std::collections::HashMap;
use std::fs;
use std::io::Write;
use colored::*;

type PlanetMap = HashMap<String, (f64, f64)>; // (L0, n)

fn planet_data() -> PlanetMap {
    let mut m = HashMap::new();
    m.insert("Mercury".to_string(), (252.2509, 4.09233445));
    m.insert("Venus".to_string(), (181.9798, 1.60213043));
    m.insert("Earth".to_string(), (100.4664, 0.98560911));
    m.insert("Mars".to_string(), (355.4530, 0.52407178));
    m.insert("Jupiter".to_string(), (34.3515, 0.08309125));
    m.insert("Saturn".to_string(), (50.0774, 0.03344482));
    m.insert("Uranus".to_string(), (314.0550, 0.01172312));
    m.insert("Neptune".to_string(), (304.3487, 0.00597733));
    m
}

const OBLIQUITY: f64 = 23.4393;

#[derive(Serialize, Deserialize)]
struct Result {
    planet: String,
    ra_deg: f64,
    dec_deg: f64,
    ra: String,
    dec: String,
}

fn ecliptic_to_equatorial(lon: f64, lat: f64, obl: f64) -> (f64, f64) {
    let lon_rad = lon.to_radians();
    let lat_rad = lat.to_radians();
    let obl_rad = obl.to_radians();
    let ra = (lon_rad.sin() * obl_rad.cos() - lat_rad.tan() * obl_rad.sin()).atan2(lon_rad.cos());
    let dec = (lat_rad.sin() * obl_rad.cos() + lat_rad.cos() * obl_rad.sin() * lon_rad.sin()).asin();
    let ra_deg = (ra.to_degrees() % 360.0 + 360.0) % 360.0;
    let dec_deg = dec.to_degrees();
    (ra_deg, dec_deg)
}

fn ra_to_hms(ra_deg: f64) -> String {
    let hours = ra_deg / 15.0;
    let h = hours.floor() as i32;
    let m = ((hours - h as f64) * 60.0).floor() as i32;
    let s = (hours - h as f64 - m as f64 / 60.0) * 3600.0;
    format!("{:02}h {:02}m {:05.2}s", h, m, s)
}

fn dec_to_dms(dec_deg: f64) -> String {
    let sign = if dec_deg >= 0.0 { '+' } else { '-' };
    let dec_abs = dec_deg.abs();
    let d = dec_abs.floor() as i32;
    let m = ((dec_abs - d as f64) * 60.0).floor() as i32;
    let s = (dec_abs - d as f64 - m as f64 / 60.0) * 3600.0;
    format!("{}{:02}° {:02}' {:05.2}\"", sign, d, m, s)
}

fn compute_position(date: DateTime<Utc>, planet: &str, data: &PlanetMap) -> (f64, f64) {
    let j2000 = Utc.with_ymd_and_hms(2000, 1, 1, 12, 0, 0).unwrap();
    let delta = (date - j2000).num_seconds() as f64 / 86400.0;
    let (l0, n) = data[planet];
    let lon = (l0 + n * delta) % 360.0;
    ecliptic_to_equatorial(lon, 0.0, OBLIQUITY)
}

fn main() {
    let matches = App::new("Planet Finder")
        .arg(Arg::with_name("date").long("date").takes_value(true).help("Дата (YYYY-MM-DD)"))
        .arg(Arg::with_name("planet").long("planet").takes_value(true).help("Название планеты"))
        .arg(Arg::with_name("list").long("list").help("Показать все планеты"))
        .arg(Arg::with_name("export-json").long("export-json").takes_value(true).help("Экспорт в JSON"))
        .arg(Arg::with_name("export-csv").long("export-csv").takes_value(true).help("Экспорт в CSV"))
        .get_matches();

    let date_str = matches.value_of("date").unwrap_or("");
    let date = if date_str.is_empty() {
        Utc::now()
    } else {
        let naive = NaiveDate::parse_from_str(date_str, "%Y-%m-%d").expect("Неверный формат даты");
        Utc.from_utc_datetime(&naive.and_hms_opt(0,0,0).unwrap())
    };

    let planet_name = matches.value_of("planet");
    let planets = if let Some(p) = planet_name {
        vec![p.to_string()]
    } else {
        planet_data().keys().cloned().collect::<Vec<_>>()
    };

    let data = planet_data();
    let mut results = Vec::new();
    for p in planets {
        let (ra, dec) = compute_position(date, &p, &data);
        results.push(Result {
            planet: p.clone(),
            ra_deg: ra,
            dec_deg: dec,
            ra: ra_to_hms(ra),
            dec: dec_to_dms(dec),
        });
    }

    if let Some(file) = matches.value_of("export-json") {
        let json = serde_json::to_string_pretty(&results).unwrap();
        fs::write(file, json).unwrap();
        println!("{}", format!("Экспортировано в {} (JSON)", file).green());
    } else if let Some(file) = matches.value_of("export-csv") {
        let mut wtr = csv::Writer::from_path(file).unwrap();
        wtr.write_record(&["planet", "ra_deg", "dec_deg", "ra", "dec"]).unwrap();
        for r in &results {
            wtr.write_record(&[&r.planet, &r.ra_deg.to_string(), &r.dec_deg.to_string(), &r.ra, &r.dec]).unwrap();
        }
        wtr.flush().unwrap();
        println!("{}", format!("Экспортировано в {} (CSV)", file).green());
    } else {
        println!("{}", format!("Положение планет на {}:\n", date.format("%Y-%m-%d")).cyan());
        for r in results {
            println!("{}:", r.planet.yellow());
            println!("  RA  = {}  ({:.2}°)", r.ra, r.ra_deg);
            println!("  Dec = {}  ({:.2}°)", r.dec, r.dec_deg);
            println!();
        }
    }
}
